test/test.jl

```diff
@@ -110,3 +109,110 @@ end
 
 # Test @test_approx_eq_eps
 # TODO
+
+ts = @testset "@testset should return the testset" begin
+    @test true
+end
+@test typeof(ts) == Base.Test.DefaultTestSet
+@test typeof(ts.results[1]) == Base.Test.Pass
+
+tss = @testloop "@testloop should return an array of testsets: $i" for i in 1:3
+    @test true
+end
+@test length(tss) == 3
+@test typeof(tss[1]) == Base.Test.DefaultTestSet
+@test typeof(tss[1].results[1]) == Base.Test.Pass
+
+# now we're done running tests with DefaultTestSet so we can go back to STDOUT
+redirect_stdout(OLD_STDOUT)
+
+# import the methods needed for defining our own testset type
+import Base.Test: record, finish, get_testset_depth, get_testset
```

Comment:
I guess to be clear, `record` and `finish` are the things you need to implement, but you will almost surely want to use `get_testset_depth` and `get_testset`. I panicked when I first read the PR because I couldn't remember what those last two did, and thought I'd leaked something bad in the interface!